pkg/controller/direct/compute/firewallpolicyrule/firewallpolicyrule_controller.go

```diff
@@ -0,0 +1,297 @@
+}
+
+func (m *firewallPolicyRuleModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.ComputeFirewallPolicyRule{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Set label managed-by-cnrm: true
+	obj.ObjectMeta.Labels["managed-by-cnrm"] = "true"
+
+	// Get firewall policy
+	firewallPolicyRef, err := ResolveComputeFirewallPolicy(ctx, reader, obj, obj.Spec.FirewallPolicyRef)
+	if err != nil {
+		return nil, err
+
+	}
+	obj.Spec.FirewallPolicyRef.External = firewallPolicyRef.External
+	firewallPolicy := obj.Spec.FirewallPolicyRef.External
+
+	// Get priority
+	priority := obj.Spec.Priority
+
+	firewallPolicyRuleAdapter := &firewallPolicyRuleAdapter{
+		firewallPolicy: firewallPolicy,
+		priority:       priority,
+		desired:        obj,
+		reader:         reader,
+	}
+
+	// Get GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("building gcp client: %w", err)
+	}
+	firewallPolicyRuleAdapter.firewallPoliciesClient = gcpClient
+
+	return firewallPolicyRuleAdapter, nil
+}
+
+func (m *firewallPolicyRuleModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+func (a *firewallPolicyRuleAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting ComputeFirewallPolicyRule", "priority", a.priority)
+
+	firewallPolicyRule, err := a.get(ctx)
+	if err != nil {
+		// When a certain rule does not exist, the error has code 400(bad request) instead of 404(not found)
+		// example error message:
+		// "Invalid value for field 'priority': '9000'. The firewall policy does not contain a rule at priority 9000.",
+		if direct.IsBadRequest(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+	a.actual = firewallPolicyRule
+	return true, nil
+}
+
+func (a *firewallPolicyRuleAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	var err error
+
+	err = resolveDependencies(ctx, a.reader, a.desired)
+	if err != nil {
+		return err
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating ComputeFirewallPolicyRule", "priority", a.priority)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+
+	firewallPolicyRule := ComputeFirewallPolicyRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &computepb.AddRuleFirewallPolicyRequest{
+		FirewallPolicyRuleResource: firewallPolicyRule,
+		FirewallPolicy:             a.firewallPolicy,
+	}
+	op, err := a.firewallPoliciesClient.AddRule(ctx, req)
+
+	if err != nil {
+		return fmt.Errorf("creating ComputeFirewallPolicyRule %d: %w", a.priority, err)
```

Comment:
If we start using `a.id` I think this will get clearer (I _think_ the full id is actually going to be project + firewallRule + priority)